[FAILING Test] UT TestSearchDisk fails when multipath device mapper device is found

### Which jobs are failing?

k8s.io/kubernetes/pkg/volume: fc TestSearchDisk at:
https://prow.ppc64le-cloud.cis.ibm.net/view/gs/ppc64le-kubernetes/logs/periodic-kubernetes-unit-tests-non-root-ppc64le/1860941588411191296


### Which tests are failing?

`k8s.io/kubernetes/pkg/volume: fc TestSearchDisk`

### Since when has it been failing?

`Nov 22, 2024`

### Testgrid link

_No response_

### Reason for failure (if possible)

Test is failing in envs where multipath device mapper is configured.

```
[root@debugging  kubernetes]# go test -run ^TestSearchDisk$ k8s.io/kubernetes/pkg/volume/fc
warning: GOPATH set to GOROOT (/usr/local/go) has no effect
I1125 12:49:05.934534  155946 fc_util.go:74] fc: find disk: /dev/sda, dm: /dev/dm-0, fc path: pci-0000:41:00.0-fc-0x500a0981891b8dc5-lun-0
I1125 12:49:05.934710  155946 fc_util.go:74] fc: find disk: /dev/sdb, dm: /dev/dm-0, fc path: pci-0000:41:00.0-fc-0x500a0981891b8dc5-lun-1
I1125 12:49:05.934846  155946 fc_util.go:74] fc: find disk: /dev/sdc, dm: /dev/dm-0, fc path: fc-0x5005076810213b32-lun-2
--- FAIL: TestSearchDisk (0.00s)
    --- FAIL: TestSearchDisk/PCI_disk_0 (0.00s)
        fc_util_test.go:181: matching wrong disk, expected: /dev/sda, actual: /dev/dm-0
    --- FAIL: TestSearchDisk/PCI_disk_1 (0.00s)
        fc_util_test.go:181: matching wrong disk, expected: /dev/sdb, actual: /dev/dm-0
    --- FAIL: TestSearchDisk/Non_PCI_disk (0.00s)
        fc_util_test.go:181: matching wrong disk, expected: /dev/sdc, actual: /dev/dm-0
FAIL
FAIL	k8s.io/kubernetes/pkg/volume/fc	0.018s
FAIL
```

### Anything else we need to know?

This is reproduced on `ppc64le` arch.

### Relevant SIG(s)

/sig storage

--- a/pkg/volume/fc/fc_util_linux_test.go
+++ b/pkg/volume/fc/fc_util_linux_test.go
@@ -1,0 +1,72 @@
+//go:build linux
+// +build linux
+
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fc
+
+import (
+	"testing"
+
+	"k8s.io/kubernetes/pkg/volume/util"
+)
+
+func TestSearchDiskMultipathDevice(t *testing.T) {
+	tests := []struct {
+		name        string
+		wwns        []string
+		lun         string
+		expectError bool
+	}{
+		{
+			name: "Non PCI disk 0",
+			wwns: []string{"500507681021a537"},
+			lun:  "0",
+		},
+		{
+			name: "Non PCI disk 1",
+			wwns: []string{"500507681022a554"},
+			lun:  "2",
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			fakeMounter := fcDiskMounter{
+				fcDisk: &fcDisk{
+					wwns: test.wwns,
+					lun:  test.lun,
+					io:   &fakeIOHandler{},
+				},
+				deviceUtil: util.NewDeviceHandler(&fakeIOHandler{}),
+			}
+			devicePath, err := searchDisk(fakeMounter)
+			if test.expectError && err == nil {
+				t.Errorf("expected error but got none")
+			}
+			if !test.expectError && err != nil {
+				t.Errorf("got unexpected error: %s", err)
+			}
+			// if no disk matches input wwn and lun, exit
+			if devicePath == "" && !test.expectError {
+				t.Errorf("no fc disk found")
+			}
+			if devicePath != "/dev/dm-1" {
+				t.Errorf("multipath device not found dm-1 expected got [%s]", devicePath)
+			}
+		})
+	}
+}

--- a/pkg/volume/fc/fc_util_test.go
+++ b/pkg/volume/fc/fc_util_test.go
@@ -17,6 +17,7 @@ limitations under the License.
 package fc
 
 import (
+	"errors"
 	"os"
 	"reflect"
 	"testing"
@@ -75,7 +76,13 @@ func (handler *fakeIOHandler) ReadDir(dirname string) ([]os.FileInfo, error) {
 		f6 := &fakeFileInfo{
 			name: "fc-0x5005076810213b32-lun-25",
 		}
-		return []os.FileInfo{f4, f5, f6, f1, f2, f3}, nil
+		f7 := &fakeFileInfo{
+			name: "fc-0x500507681021a537-lun-0",
+		}
+		f8 := &fakeFileInfo{
+			name: "fc-0x500507681022a554-lun-2",
+		}
+		return []os.FileInfo{f4, f5, f6, f1, f2, f3, f7, f8}, nil
 	case "/sys/block/":
 		f := &fakeFileInfo{
 			name: "dm-1",
@@ -91,7 +98,15 @@ func (handler *fakeIOHandler) ReadDir(dirname string) ([]os.FileInfo, error) {
 }
 
 func (handler *fakeIOHandler) Lstat(name string) (os.FileInfo, error) {
-	return nil, nil
+	links := map[string]string{
+		"/sys/block/dm-1/slaves/sde": "sde",
+		"/sys/block/dm-1/slaves/sdf": "sdf",
+		"/sys/block/dm-1/slaves/sdg": "sdg",
+	}
+	if dev, ok := links[name]; ok {
+		return &fakeFileInfo{name: dev}, nil
+	}
+	return nil, errors.New("device not found for mock")
 }
 
 func (handler *fakeIOHandler) EvalSymlinks(path string) (string, error) {
@@ -108,12 +123,24 @@ func (handler *fakeIOHandler) EvalSymlinks(path string) (string, error) {
 		return "/dev/sdx", nil
 	case "/dev/disk/by-id/scsi-3600508b400105e210000900000490000":
 		return "/dev/sdd", nil
+	case "/dev/disk/by-path/fc-0x500507681021a537-lun-0":
+		return "/dev/sde", nil
+	case "/dev/disk/by-path/fc-0x500507681022a554-lun-2":
+		return "/dev/sdf", nil
+	case "/dev/sde":
+		return "/dev/sde", nil
+	case "/dev/sdf":
+		return "/dev/sdf", nil
 	}
 	return "", nil
 }
 
 func (handler *fakeIOHandler) WriteFile(filename string, data []byte, perm os.FileMode) error {
 	return nil
+}
+
+func (handler *fakeIOHandler) ReadFile(filename string) ([]byte, error) {
+	return nil, nil
 }
 
 func TestSearchDisk(t *testing.T) {
@@ -164,7 +191,7 @@ func TestSearchDisk(t *testing.T) {
 					lun:  test.lun,
 					io:   &fakeIOHandler{},
 				},
-				deviceUtil: util.NewDeviceHandler(util.NewIOHandler()),
+				deviceUtil: util.NewDeviceHandler(&fakeIOHandler{}),
 			}
 			devicePath, err := searchDisk(fakeMounter)
 			if test.expectError && err == nil {

